sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceEventClient.java

```diff
@@ -0,0 +1,402 @@
+                        getEntry(unexpectedResources, unexpectedInOffer.getOffer())
+                                .addAll(unexpectedInOffer.getResources());
+                    }
+                    break;
+                }
+            }
+        }
+
+        // Return the combined listing of unexpected resources across all services:
+        return anyFailedClients
+                ? UnexpectedResourcesResponse.failed(unexpectedResources.values())
+                : UnexpectedResourcesResponse.processed(unexpectedResources.values());
+    }
+
+    /**
+     * Maps the provided status to the service that owns its task, then queries that service with the status.
+     *
+     * <p>This is an optimization which avoids querying services about task statuses that don't relate to them.
+     * <p>In addition to reducing unnecessary queries, this also improves isolation between services. They only see
+     * task statuses which relate to them.
+     */
+    @Override
+    public StatusResponse status(Protos.TaskStatus status) {
+        Optional<AbstractScheduler> service = multiServiceManager.getMatchingService(status);
+        if (!service.isPresent()) {
+            // Unrecognized service. Status for old task?
+            LOGGER.info("Received status for unknown task {}: {}",
+                    status.getTaskId().getValue(), TextFormat.shortDebugString(status));
+            return StatusResponse.unknownTask();
+        }
+        LOGGER.info("Received status for task {}: {}", status.getTaskId().getValue(), status.getState());
+        return service.get().status(status);
+    }
+
+    /**
+     * Returns a set of queues-specific endpoints to be served by the scheduler. This effectively overrides the
+     * underlying per-service endpoints with multiservice-aware versions.
+     */
+    @Override
+    public Collection<Object> getHTTPEndpoints() {
+        Collection<PlanManager> planManagers = uninstallPlan.isPresent()
+                ? Collections.singletonList(DefaultPlanManager.createProceeding(uninstallPlan.get()))
+                : Collections.emptyList(); // ... any plans to show when running normally?
+        List<Object> endpoints = new ArrayList<>();
+        endpoints.addAll(Arrays.asList(
+                new HealthResource(planManagers),
+                new PlansResource(planManagers),
+                new MultiArtifactResource(multiServiceManager),
+                new MultiConfigResource(multiServiceManager),
+                new MultiEndpointsResource(frameworkName, multiServiceManager, schedulerConfig),
+                new MultiPlansResource(multiServiceManager),
+                new MultiPodResource(multiServiceManager),
+                new MultiStateResource(multiServiceManager, new StringPropertyDeserializer())));
+        endpoints.addAll(customEndpoints);
+        return endpoints;
+    }
+
+    /**
+     * Finds the requested entry in the provided map, initializing the entry if needed.
```

Comment:
This comment could be better.  I don't really know what an `entry` is here.

Reply:
Done